lib/matplotlib/axes/_axes.py

```diff
@@ -3991,7 +3991,13 @@ def bxp(self, bxpstats, positions=None, widths=None, vert=True,
         if zorder is None:
             zorder = mlines.Line2D.zorder
 
-        zdelta = 0.1
+        # Use 'butt' as the default capstyle to avoid a visual overlap of
+        # the median line and the boxplot
+        if medianprops.get('solid_capstyle') is None:
+            medianprops['solid_capstyle'] = 'butt'
+
+        # put the mean and median line below the box for visual improvement
+        zdelta = -0.1
```

Comment:
Won't this lead to that the lines disappear if the box is colored?

Reply:
You are right. The lines disappear due to the change of the zdelta. I am not sure what the intended result should be. As you described (https://github.com/matplotlib/matplotlib/issues/19409#issuecomment-1146622193), the mean and median lines have a different cap style. How should the desired result look?

Reply:
My comment there did clearly not take filled boxes into account...

I think that is really the main problem here: what should it look like? I've googled a bit and there seems to be no real consensus if the lines should extend in a certain way.

Anyway, I guess that there are two options here:

1. simply change the caps as you currently do and do not change the zorder.
2. use something like `clip_on=True, clip_path=box` to the line (where `box` is the actual box object). If that is the right one is hard to say though. This should at least work for the median line. (The mean line may disappear if it is not in the box though.) However, I think that this will give exactly the same result as it is clipped using the "theoretical zero width line" rather than the actual line.

So probably just reverting the zorder change is enough.

If you can provide an image of the results (e..g using any of the examples in #19409, possibly with exaggerated line widths), that would simplify the review. (Just paste it in a comment.)

Reply:
I would actually strongly consider an image test here so we can codify what this should look like.

Reply:
Thank you for the detailed answer. I will go with option 1 then. 
Also I will read further into the image tests and provide a test for it.